### ISSUE TITLE
Zed fails to launch on X11 under tmux session started at wayland.

### ISSUE THREAD

**Issue by AlvaroParker** [zed-industries/zed#17199]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

When you create a `tmux` session on wayland and then log into an X11 window manager (for example `i3`) and attach to that `tmux` session, zed will fail to load because wayland env vars are still set (`WAYLAND_DISPLAY`). I do not know if this is applicable as a bug because, after all, I'm passing the wrong env vars to zed. 

To reproduce: 
- Start a wayland session (`sway`, `hyprland`, etc)
- Open a terminal and type `tmux`
- Detach from the `tmux` session (`C-b d`) 
- Logout fromthe wayland session
- Start an X11 session (Like `i3`)
- Open a terminal
- Type `tmux attach -t 0` (assuming the tmux session you created on wayland was the first session)
- Run zed: `zeditor --foreground` 

### Environment

Zed: v0.150.4 (Zed)
OS: Linux X11 arch unknown
Memory: 6.7 GiB
Architecture: x86_64
GPU: AMD Radeon Vega 8 Graphics (RADV RAVEN) || radv || Mesa 24.2.1-arch1.1

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.

<details><summary>Zed.log</summary><pre>

</pre></details>

**Comment by notpeter** [comment]
Yeah, that's a bit wonky.  I don't think we want to try and be "smart": ignoring the environment variables a user has set and trying to discover the "real" session (principle of least surprise).

I don't immediately have a suggestion for how to update your tmux session to match the currently running desktop session. Perhaps you could have a script which used `loginctl show-session 2 -p Type` and then unset/reset the appropriate variables?

Just curious how we fail in this scenario -- do we gracefully say "couldn't start" or some such or do half-start and leave an orphan process which needs to be killed. If it's the latter, we would welcome a PR that would at least make Zed fail gracefully.

**Comment by AlvaroParker** [comment]
> Just curious how we fail in this scenario -- do we gracefully say "couldn't start" or some such? Or do half-start and leave an orphan process which needs to be killed? Or worse, panic? We would welcome a PR that would at least make Zed fail gracefully.

@notpeter  Currently zed panics: 
```
thread 'main' panicked at crates/gpui/src/platform/linux/wayland/client.rs:409:49:
called `Result::unwrap()` on an `Err` value: NoCompositor
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
On line `409` (First line of `WaylandClient::new`) : 
```rust
let conn = Connection::connect_to_env().unwrap();
```
Which is pretty much the first function call when creating a new `gpui` app on wayland using `App::new()` so doing
```rust
        let conn = match Connection::connect_to_env() {
            Ok(c) => c,
            Err(err) => {
                log::error!("Failed to connect to Wayland server: {}", err);
                std::process::exit(1);
            }
        };
```
Should be enough. Same goes for `x11`. I can send a PR if this change is acceptable on the `gpui` create.
